### PR TITLE
[SPARK-15215] [SQL] Fix Explain Parsing and Output

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

This PR is to address a few existing issues in `EXPLAIN`:
- The `EXPLAIN` options `LOGICAL | FORMATTED | EXTENDED | CODEGEN` should not be 0 or more match. It should 0 or one match. Parser does not allow users to use more than one option in a single command.
- The option `LOGICAL` is not supported. Issue an exception when users specify this option in the command.
- The output of `EXPLAIN` contains a weird empty line when the output of analyzed plan is empty. We should remove it. For example:
  
  ```
  == Parsed Logical Plan ==
  CreateTable CatalogTable(`t`,CatalogTableType(MANAGED),CatalogStorageFormat(None,Some(org.apache.hadoop.mapred.TextInputFormat),Some(org.apache.hadoop.hive.ql.io.  HiveIgnoreKeyTextOutputFormat),None,false,Map()),List(CatalogColumn(col,int,true,None)),List(),List(),List(),-1,,1462725171656,-1,Map(),None,None,None), false
  
  == Analyzed Logical Plan ==
  
  CreateTable CatalogTable(`t`,CatalogTableType(MANAGED),CatalogStorageFormat(None,Some(org.apache.hadoop.mapred.TextInputFormat),Some(org.apache.hadoop.hive.ql.io.  HiveIgnoreKeyTextOutputFormat),None,false,Map()),List(CatalogColumn(col,int,true,None)),List(),List(),List(),-1,,1462725171656,-1,Map(),None,None,None), false
  
  == Optimized Logical Plan ==
  CreateTable CatalogTable(`t`,CatalogTableType(MANAGED),CatalogStorageFormat(None,Some(org.apache.hadoop.mapred.TextInputFormat),Some(org.apache.hadoop.hive.ql.io.  HiveIgnoreKeyTextOutputFormat),None,false,Map()),List(CatalogColumn(col,int,true,None)),List(),List(),List(),-1,,1462725171656,-1,Map(),None,None,None), false
  ...
  ```
#### How was this patch tested?

Added and modified a few test cases

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @hvanhovell Could you please review this? Thanks!

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -92,7 +92,7 @@ statement
     | CREATE TEMPORARY? FUNCTION qualifiedName AS className=STRING
         (USING resource (',' resource)*)?                              #createFunction
     | DROP TEMPORARY? FUNCTION (IF EXISTS)? qualifiedName              #dropFunction
-    | EXPLAIN explainOption* statement                                 #explain
+    | EXPLAIN (LOGICAL | FORMATTED | EXTENDED | CODEGEN)? statement    #explain
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: What does `LOGICAL` actually show?

**Comment by hvanhovell** [comment]
LGTM

cc @rxin should we also add this to 2.0?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: : ) Based on my understanding, it outputs optimized logical plan. https://issues.apache.org/jira/browse/HIVE-4812 Thus, I assume we do not need to support `Explain Logical`, which is already included in our current `Explain Extended`. Let me know if we want to support it. This is an undocumented Hive feature. 

Compared with Hive, we are not following what Hive does. For example, `Explain Extended` only outputs the AST and Physical plan.

Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: FYI, 

**Case 1: Explain without any option:**

```
hive> explain select * from t1;
OK
STAGE DEPENDENCIES:
  Stage-0 is a root stage

STAGE PLANS:
  Stage: Stage-0
    Fetch Operator
      limit: -1
      Processor Tree:
        TableScan
          alias: t1
          Statistics: Num rows: 1 Data size: 0 Basic stats: PARTIAL Column stats: NONE
          Select Operator
            expressions: col (type: int)
            outputColumnNames: _col0
            Statistics: Num rows: 1 Data size: 0 Basic stats: PARTIAL Column stats: NONE
            ListSink

Time taken: 0.756 seconds, Fetched: 17 row(s)
```

Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: **Case 2: Explain with an option Extended:**

```
hive> explain extended select * from t1;
OK
ABSTRACT SYNTAX TREE:
  
TOK_QUERY
   TOK_FROM
      TOK_TABREF
         TOK_TABNAME
            t1
   TOK_INSERT
      TOK_DESTINATION
         TOK_DIR
            TOK_TMP_FILE
      TOK_SELECT
         TOK_SELEXPR
            TOK_ALLCOLREF


STAGE DEPENDENCIES:
  Stage-0 is a root stage

STAGE PLANS:
  Stage: Stage-0
    Fetch Operator
      limit: -1
      Processor Tree:
        TableScan
          alias: t1
          Statistics: Num rows: 1 Data size: 0 Basic stats: PARTIAL Column stats: NONE
          GatherStats: false
          Select Operator
            expressions: col (type: int)
            outputColumnNames: _col0
            Statistics: Num rows: 1 Data size: 0 Basic stats: PARTIAL Column stats: NONE
            ListSink

Time taken: 0.039 seconds, Fetched: 34 row(s)
```

Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: **Case 3: Explain with an option Logical:**
```
hive> explain logical select * from t1;
OK
ABSTRACT SYNTAX TREE:
  
TOK_QUERY
   TOK_FROM
      TOK_TABREF
         TOK_TABNAME
            t1
   TOK_INSERT
      TOK_DESTINATION
         TOK_DIR
            TOK_TMP_FILE
      TOK_SELECT
         TOK_SELEXPR
            TOK_ALLCOLREF


LOGICAL PLAN:
t1 
  TableScan (TS_0)
    alias: t1
    Statistics: Num rows: 1 Data size: 0 Basic stats: PARTIAL Column stats: NONE
    Select Operator (SEL_1)
      expressions: col (type: int)
      outputColumnNames: _col0
      Statistics: Num rows: 1 Data size: 0 Basic stats: PARTIAL Column stats: NONE
      ListSink (OP_3)

Time taken: 0.041 seconds, Fetched: 27 row(s)
```

Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: **Case 4: Explain with an option Formatted:**

```
hive> explain formatted select * from t1;
OK
{"STAGE PLANS":{"Stage-0":{"Fetch Operator":{"limit:":"-1","Processor Tree:":{"TableScan":{"alias:":"t1","children":{"Select Operator":{"expressions:":"col (type: int)","outputColumnNames:":["_col0"],"children":{"ListSink":{}},"Statistics:":"Num rows: 1 Data size: 0 Basic stats: PARTIAL Column stats: NONE"}},"Statistics:":"Num rows: 1 Data size: 0 Basic stats: PARTIAL Column stats: NONE"}}}}},"STAGE DEPENDENCIES":{"Stage-0":{"ROOT STAGE":"TRUE"}}}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 2]
```diff
@@ -212,8 +212,7 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
     s"""== Parsed Logical Plan ==
        |${stringOrError(logical)}
        |== Analyzed Logical Plan ==
-       |${stringOrError(output)}
-       |${stringOrError(analyzed)}
+       |${Seq(stringOrError(output), stringOrError(analyzed)).filter(_.nonEmpty).mkString("\n")}
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: this is fairly awkward. can you just break them into two statements.

**Comment by rxin** [comment]
Looks alright to add.

@gatorsmile can you submit a separate pr to fix the toString for CatalogTableType

It reads really really weird with all the None, Some, Option, List, etc.

**Comment by rxin** [comment]
And of course the timestamp 1462725171656

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Let me break them now. Thanks!

**Comment by gatorsmile** [comment]
Sure, will submit a separate PR for it.

**Comment by rxin** [comment]
LGTM pending Jenkins.

**Comment by rxin** [comment]
BTW do we need to update the description? Is it still accurate?

**Comment by gatorsmile** [comment]
Just updated it  and now I think it is more readable. Let me know if anything is not clear. Thanks!

**Comment by rxin** [comment]
@hvanhovell feel free to merge.

**Pull request closed without merging**